graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala

```diff
@@ -0,0 +1,138 @@
+  val LABEL_COLUMN_PREFIX = ":"
+}
+
+/**
+ * The entry point for using property graphs in Spark.
+ *
+ * Provides factory methods for creating [[PropertyGraph]] instances.
+ *
+ * Wraps a [[org.apache.spark.sql.SparkSession]].
+ *
+ * @since 3.0.0
+ */
+trait CypherSession {
+
+  def sparkSession: SparkSession
+
+  /**
+   * Creates a [[PropertyGraph]] from a sequence of [[NodeFrame]]s and [[RelationshipFrame]]s.
+   * At least one [[NodeFrame]] has to be provided.
+   *
+   * For each label set and relationship type there can be at most one [[NodeFrame]] and at most one
+   * [[RelationshipFrame]], respectively.
+   *
+   * @param nodes         NodeFrames that define the nodes in the graph
+   * @param relationships RelationshipFrames that define the relationships in the graph
+   * @since 3.0.0
+   */
+  def createGraph(nodes: Seq[NodeFrame], relationships: Seq[RelationshipFrame]): PropertyGraph
+
+  /**
+   * Creates a [[PropertyGraph]] from a sequence of [[NodeFrame]]s and [[RelationshipFrame]]s.
+   * At least one [[NodeFrame]] has to be provided.
+   *
+   * For each label set and relationship type there can be at most one [[NodeFrame]] and at most one
+   * [[RelationshipFrame]], respectively.
+   *
+   * @param nodes         NodeFrames that define the nodes in the graph
+   * @param relationships RelationshipFrames that define the relationships in the graph
+   * @since 3.0.0
+   */
+  def createGraph(
+      nodes: java.util.List[NodeFrame],
+      relationships: java.util.List[RelationshipFrame]): PropertyGraph = {
+    createGraph(JavaConverters.asScalaBuffer(nodes), JavaConverters.asScalaBuffer(relationships))
+  }
+
+  /**
+   * Creates a [[PropertyGraph]] from nodes and relationships.
+   *
+   * The given DataFrames need to adhere to the following column naming conventions:
+   *
+   * {{{
+   *     Id column:        `$ID`            (nodes and relationships)
+   *     SourceId column:  `$SOURCE_ID`     (relationships)
+   *     TargetId column:  `$TARGET_ID`     (relationships)
+   *
+   *     Label columns:    `:{LABEL_NAME}`  (nodes)
+   *     RelType columns:  `:{REL_TYPE}`    (relationships)
+   *
+   *     Property columns: `{Property_Key}` (nodes and relationships)
+   * }}}
+   *
+   * @see [[CypherSession]]
+   * @param nodes         node DataFrame
+   * @param relationships relationship DataFrame
+   * @since 3.0.0
+   */
+  def createGraph(nodes: DataFrame, relationships: DataFrame): PropertyGraph = {
+    val idColumn = CypherSession.ID_COLUMN
+    val sourceIdColumn = CypherSession.SOURCE_ID_COLUMN
+    val targetIdColumn = CypherSession.TARGET_ID_COLUMN
+
+    val labelColumns = nodes.columns.filter(_.startsWith(CypherSession.LABEL_COLUMN_PREFIX)).toSet
+    val nodeProperties = (nodes.columns.toSet - idColumn -- labelColumns)
+      .map(col => col -> col)
+      .toMap
+
+    val trueLit = functions.lit(true)
+    val falseLit = functions.lit(false)
+
+    val labelSets = labelColumns.subsets().toSet + Set.empty
```

Comment:
The `subset()` cause running time explosion.
Suppose we have `N` labels, then it will generate `2^N` subsets,
then you will scan and filter the input `nodes` dataframe `2^N` times (the worst case the `nodes` dataframe is not cached).

Reply:
Each of those sets represents a possible label combination (the empty set is the case where a node has no label). The input dataframe has no information about the possible label combinations. From looking at its schema, you can just derive which labels exist and we have to assume that any combination of those labels exists in the data. The alternative would be to look at the data, which we want to avoid.

Generally, this method is just for convenience. To avoid the combinatorical problem, one can use `NodeFrame` and `RelationshipFrame` for each possible label combination / relationship type.

Do you see any other alternative? We could however, make that more clear in the documentation of that method. Wdyt?

Reply:
Yes, logically we should do this, but we can still do some optimization, such as we do a `partitionBy` to shuffle data to different partitions, and then extract each partition to be a result dataframe, we could discuss them later when the optimization needed.

Reply:
I created https://issues.apache.org/jira/browse/SPARK-28513